events.cpp

```diff
@@ -233,3 +234,104 @@ std::string Events::Write(std::string worldName, std::string month, int year) {
 
     return text;
 }
+
+//----------------------------------------------------------------------------
+void StrEvent::writeJson(JsonReport &of, ARegionList &) const
+{
+	of.PutStr(str_.c_str());
+}
+
+//----------------------------------------------------------------------------
+TaxEvent::TaxEvent(const std::string &uname, int amt, ARegion *reg)
+: uname_(uname)
+, amt_(amt)
+, reg_(reg)
+{
+}
+
+std::string TaxEvent::str(ARegionList &regions) const
+{
+	std::ostringstream os;
+	os << uname_ << ": Collects $" << amt_ << " in taxes in ";
+	os << reg_->ShortPrint(&regions) << '.';
+	return os.str();
+}
+
+void TaxEvent::writeJson(JsonReport &of, ARegionList &regions) const
+{
+	of.StartDict(NULL);
+	of.PutPairStr("type", "Tax");
+	of.PutPairStr("name", uname_.c_str());
+	of.PutPairInt("amount", amt_);
+
+	of.StartDict("region");
+	reg_->CPrint(&of, &regions, false);
+	of.EndDict();
+
+	of.EndDict();
+}
+
+//----------------------------------------------------------------------------
+GiveEvent::GiveEvent(const std::string &gname, int inum, int amt, const std::string &tname, bool reversed)
+: gname_(gname)
+, inum_(inum)
+, amt_(amt)
+, tname_(tname)
+, reversed_(reversed)
+{
+}
+
+std::string GiveEvent::str(ARegionList &) const
+{
+	std::ostringstream os;
+	if (reversed_)
+		os << tname_ << ": Receives " << ItemString(inum_, amt_) << " from " << gname_ << '.';
+	else
+		os << gname_ << ": Gives " << ItemString(inum_, amt_) << " to " << tname_ << '.';
+
+	return os.str();
+}
+
+void GiveEvent::writeJson(JsonReport &of, ARegionList &regions) const
+{
+	of.StartDict(NULL);
+	of.PutPairStr("type", reversed_ ? "Receive" : "Give");
+	of.PutPairStr("giver", gname_.c_str());
+	of.PutPairStr("receiver", tname_.c_str());
+	of.PutPairStr("abbr", ItemDefs[inum_].abr);
+	of.PutPairInt("amount", amt_);
+	of.EndDict();
+}
+
+//----------------------------------------------------------------------------
+ProduceEvent::ProduceEvent(const std::string &uname, int inum, int amt, ARegion *reg)
+: uname_(uname)
+, inum_(inum)
+, amt_(amt)
+, reg_(reg)
+{
+}
+
+std::string ProduceEvent::str(ARegionList &regions) const
+{
+	std::ostringstream os;
+	os << uname_ << ": Produces " << ItemString(inum_, amt_) << " in " << reg_->ShortPrint(&regions) << '.';
```

Comment:
we mostly using string concat, not piping, can we use most common Atlantis code style to keep it consistent?

Reply:
Concatenation blows up the code and is not efficient as each time strings are copied. This is standart style in C/C++

Reply:
My point is that code consistency is more important than efficiency. And to be honest, efficiency at Atlantis is not an issue.

Reply:
Writing 2x code and spending 2x time for that cannot be adjusted just because it was no used before. This is super standard C code.

Reply:
There is no normal string.format type in C11 and this approach converts numbers to strings without unnecessary function calls. Also it is easy to follow what happens and don't need any boilerplate.

Reply:
If yoi have worked with Java or C# then there is `StringBuilder` class, which practically does the same thing as `stringstream`.
`ostringstream` just means that you can only WRITE to it and no reading can happen, i.e. exactly what we need in this case.

Reply:
The good point for `stringstream` is that you can put in `string` or `number` and it will be handled correctly for you. No need to call additional methods or constructs.

Reply:
Yeah, it's part of modernizing the code. In Miskatonic, I've started bringing in some more modern concepts like templates and lambda functions. It makes the code a lot easier to read.